EngineLayer/ModernSearch/ModernSearchEngine.cs

```diff
@@ -39,84 +40,44 @@ protected override MetaMorpheusEngineResults RunSpecific()
-                    // this is just PRELIMINARY precursor-mass filtering
-                    // additional checks are made later to ensure that the theoretical precursor mass is acceptable
-                    IEnumerable<AllowedIntervalWithNotch> notches = MassDiffAcceptor.GetAllowedPrecursorMassIntervalsFromObservedMass(scan.PrecursorMass);
-
-                    double lowestMassPeptideToLookFor = notches.Min(p => p.AllowedInterval.Minimum);
-                    double highestMassPeptideToLookFor = notches.Max(p => p.AllowedInterval.Maximum);
+                    byte[] scoringTable = new byte[PeptideIndex.Count];
+                    List<int> idsOfPeptidesPossiblyObserved = new List<int>(PeptideIndex.Count);
 
-                    // first-pass scoring
-                    IndexedScoring(FragmentIndex, allBinsToSearch, scoringTable, byteScoreCutoff, idsOfPeptidesPossiblyObserved, scan.PrecursorMass, lowestMassPeptideToLookFor, highestMassPeptideToLookFor, PeptideIndex, MassDiffAcceptor, MaxMassThatFragmentIonScoreIsDoubled, CommonParameters.DissociationType);
-
-                    // done with indexed scoring; refine scores and create PSMs
-                    foreach (int id in idsOfPeptidesPossiblyObserved)
+                    for (int scanIndex = range.Item1; scanIndex < range.Item2; scanIndex++)
                     {
-                        PeptideWithSetModifications peptide = PeptideIndex[id];
-
-                        List<Product> peptideTheorProducts = peptide.Fragment(CommonParameters.DissociationType, FragmentationTerminus.Both).ToList();
-
-                        List<MatchedFragmentIon> matchedIons = MatchFragmentIons(scan, peptideTheorProducts, CommonParameters);
-
-                        double thisScore = CalculatePeptideScore(scan.TheScan, matchedIons);
-                        int notch = MassDiffAcceptor.Accepts(scan.PrecursorMass, peptide.MonoisotopicMass);
+                        Ms2ScanWithSpecificMass scan = ListOfSortedMs2Scans[scanIndex];
 
-                        bool meetsScoreCutoff = thisScore >= CommonParameters.ScoreCutoff;
-                        bool scoreImprovement = PeptideSpectralMatches[i] == null || (thisScore - PeptideSpectralMatches[i].RunnerUpScore) > -PeptideSpectralMatch.ToleranceForScoreDifferentiation;
+                        // do a fast rough first-pass scoring for this scan
+                        IndexScoreScan(scan, scoringTable, byteScoreCutoff, idsOfPeptidesPossiblyObserved, CommonParameters.DissociationType);
 
-                        if (meetsScoreCutoff && scoreImprovement)
+                        // take top 10 indexed-scored peptides and re-score them using the more accurate but slower scoring algorithm
+                        // NOTE: this "Take 10" section means that modern search is NOT GUARANTEED to produce the same results as classic search
```

Comment:
:(

Reply:
there are ways around this, fortunately. we could include the distance from the bin edge with each fragment in the index - so instead of just being a peptide ID number, it would be a peptide ID number and a dalton "error" from the start or end of the bin, giving an exact dalton value rather than approximation. I'm pretty sure this would be slightly faster than the current implementation. it would of course cost more hard disk space/RAM.

Reply:
Not sure about how this could be faster.

Reply:
What Shortreed did was: find the score of the 10th id, then take all the ids >= the score.

Reply:
It is possible to keep all the 10 for PEP...